jctools-experimental/src/main/java/org/jctools/uc/FlatCombining.java

```diff
@@ -0,0 +1,181 @@
+ * The only consequcne is that the {@link #apply(Operation)} method may block.
+ * <p>
+ * This is conceptually similar to wrapping the data structure in a mutex, but the {@code FlatCombining} implementation
+ * will scale positively with the level of concurrency, while contended locks typically scales negatively.
+ * Negative scalability is when the system-wide throughput drops as the number of active threads increases.
+ * <p>
+ * The Flat Combining concept is described by the paper
+ * <a href="https://www.cs.bgu.ac.il/~hendlerd/papers/flat-combining.pdf">
+ * Flat Combining and the Synchronization-Parallelism Tradeoff</a>
+ *
+ * @param <T> The type of data managed by this flat combining instance.
+ */
+public class FlatCombining<T> {
+    private final WFIStack<OpNode> ops;
+    private final AtomicBoolean lock;
+    private final T data;
+
+    public FlatCombining(T sequentialData) {
+        ops = new WFIStack<OpNode>();
+        lock = new AtomicBoolean();
+        data = sequentialData;
+    }
+
+    /**
+     * Apply the given operation on the underlying data structure. This call may block. Multiple threads can calls
+     * this at the same time. The operations will be applied with linearizable consistency.
+     * <p>
+     * The operation will only be applied once to the data structure, but there is no guarantee that it will be
+     * applied by the calling thread.
+     *
+     * @param operation The operation to apply to the underlying data structure.
+     * @param <R>       The type of return value produced by the operation, if any.
+     * @param <E>       The type of exception the operation might throw, if any.
+     * @return The result returned by the operation.
+     * @throws E                    If the operation throws an exception.
+     * @throws InterruptedException If the apply method ends up blocking but the thread is interrupted.
+     */
+    public <R, E extends Exception> R apply(Operation<T, R, E> operation) throws E, InterruptedException {
+        OpNode op = enqueue(operation);
+
+        for (; ; ) {
+            if (tryLock()) {
+                Iterable<OpNode> nodes = combineAndApplyAllOperations();
+                unlockAndUnparkWaiters(nodes);
+            } else {
+                op.await();
+            }
+            if (op.isDone()) {
+                try {
+                    return op.getResult();
+                } catch (InterruptedException ie) {
+                    throw ie;
+                } catch (Exception e) {
+                    //noinspection unchecked
+                    throw (E) e;
+                } catch (Throwable th) {
+                    throw new RuntimeException(th);
+                }
+            }
+        }
+    }
+
+    private <R, E extends Exception> OpNode enqueue(Operation<T, R, E> operation) {
+        if (operation == null) {
+            throw new NullPointerException("The operation cannot be null.");
+        }
+        OpNode op = new OpNode(operation);
```

Comment:
2 comments:
# I often see combiners using a thread local reusable node to save allocating in the hot path, do you think could worth doing it? Given that execution completed mark the end of the node lifecycle it should make possible to pool it
# is see we poll with FIFO policy, why not leaving the choice of using different data-structures (eg lifo/fifo) to store the ops?

Reply:
We could do pooling. An apply method that separates enqueueing and application in time via a Future can't rely on pooling, though. Those operations will have to be unpooled. Perhaps they need to be in a separate stack as well to prevent the pooled stack from growing too big. The lock could be turned from a boolean to an int, and include a sequence so we can periodically truncate the stack and avoid leaking.

Applying the operations in FIFO order is how the linearisable consistency is achieved. Would be easy to add an option to turn that off.

Reply:
Thinking about the consistency... since concurrent blocking apply calls overlap in time, I think we can pick an arbitrary ordering of these overlapping operations, and still be considered linearisable. The operations were already racing, so there's no externally observable sequence of events that can disagree with applying the operations in LIFO order.